### DISCUSSION TITLE
Why can't 'alpha' be found?

### DISCUSSION THREAD

**Discussion by MScKarl** [idaholab/moose#20872] [Q&A General] [answered]
Hi!
Using the code below, 

```
[Functions]
	[./alpha] 
	type = ParsedFunction
	value = abs(alpha_0)*(sgn_0-sqrt(r_0/sqrt(x*x+y*y))*cos(atan2(x,y)/2))
	#  alpha_0, sgn_0, r_0 are constants
	[../]
[]

[Materials]
        #Total Free Energy 
        [./f_eta]
        type = ParsedMaterial
        f_name = f_eta # Name of the function
        args = 'alpha beta gamma eta' #the coefficients
        function = '(alpha*eta+beta*eta^3+gamma*eta^5)' 
        [../]
[]
```

I get the following error message:
```
*** ERROR ***
f_eta: Coupled variable 'alpha' was not found
```
My question is, why can't alpha be found?

Best regards,
Karl

**Comment by GiudGiud** [comment, marked as answer]
Hello

If you look at the docs for ParsedMaterial you ll see that `args` is to declare variables, not for functions. 
There is actually not support for functions in this Material, except for parsed expressions. Since alpha is a parsed expression, you can write it directly in the ParsedMaterial.

Guillaume

**Reply by MScKarl** [reply]
Hi again,
Thank you for your answer. 
I am sorry to bother you again. But if I understood your answer correctly, I should just use alpha directly in my main equation. Like this:

[Materials]

        #Total Free Energy 
        [./f_eta]
        type = ParsedMaterial
        f_name = f_eta # Name of the function
        args = 'alpha_0 sgn_0 r_0 beta gamma eta' #the coefficients
        function = '((abs(alpha_0)*(sgn_0-sqrt(r_0/sqrt(x*x+y*y))*cos(atan2(x,y)/2)))*eta+beta*eta^3+gamma*eta^5)' 
        [../]

[]

When I tried this, I got the following error: 
*** ERROR ***
Invalid function
((abs(alpha_0)*(sgn_0-sqrt(r_0/sqrt(x*x+y*y))*cos(atan2(x,y)/2)))*eta+beta*eta^3+gamma*eta^5)
alpha_0,sgn_0,r_0,beta,gamma,eta
in ParsedMaterialHelper.
Syntax error: Unknown identifier

What am I missing?

Best regards
Karl

**Reply by GiudGiud** [reply]
hello 

you need to declare the constants for the parser to know what it refers to

constant_names and constant_expressions are the two to set
https://mooseframework.inl.gov/source/materials/ParsedMaterial.html

Guillaume

**Reply by MScKarl** [reply]
But they're declared and given an initial condition. 

[AuxVariables]

	[./alpha_0]
        order = CONSTANT
        family = MONOMIAL
        initial_condition = 1.0
	[../]

	[./sgn_0]
	order = CONSTANT
	family = MONOMIAL
	initial_condition = 1.0
	#Ändra samtida som alpha_0
	[../]

	[./beta]
        order = CONSTANT
        family = MONOMIAL
        initial_condition = 1.0
	[../]
	[./gamma]
        order = CONSTANT
        family = MONOMIAL
        initial_condition = 1.0
	[../]
	[./g] #vary
        order = CONSTANT
        family = MONOMIAL
        initial_condition = 0.001
	[../]

	[./r_0] 
        order = CONSTANT
        family = MONOMIAL
        initial_condition = 1
	[../]
[]

[Variables]

	[./eta]
	[./InitialCondition]
	type = RandomIC
	min = -1
	max = 1
	[../]
	[../]
[]

**Reply by GiudGiud** [reply]
you need to tell the parser about it. It's explicit in the parameters.

**Reply by MScKarl** [reply]
[Functions]

 	[./alpha]  #α ≡ |α_ 0| (sgn(α_ 0)-√(r_0/r)*cos(θ/2))
	type = ParsedFunction
	value = abs(alpha_0)*(sgn_0-sqrt(r_0/sqrt(x*x+y*y))*cos(atan2(x,y)/2))
	[../]
[]

[Materials]

	[./constants]
	type = GenericConstantMaterial
	prop_names = 'alpha_0 sgn_0 beta gamma r_0'
	prop_values = '1.0 1.0 1.0 1.0 1.0'
	[../]

        #Total Free Energy 
        [./f_eta]
	type = ParsedMaterial
	f_name = f_eta # Name of the function
	args = 'sgn_0 r_0 beta gamma eta' #the coefficients
	constat_names = 'alpha x y'#_0 sgn_0 r_0 x y
	function = 'alpha*eta+beta*eta^3+gamma*eta^5'
	#function = 'abs(alpha_0*(sgn_0-sqrt(r_0/(x*x+y*y))*cos(atan2(x,y)/2)))*eta+beta*eta^3+gamma*eta^5'
	[../]

[]

Why can't alpha find x and y which are coordinates from my mesh? 

Additionaly isn't possible to comined coordinates and phase variables in one function as tried in the '#function'?

**Reply by GiudGiud** [reply]
look at the parameters for parsed function. You need to declare what you are using in the parsed expression inside the ParsedFunction block.
https://mooseframework.inl.gov/source/functions/MooseParsedFunction.html

**Reply by MScKarl** [reply]
Hi again, 
Thank you. But I tried this:
[Functions]

 	[./alpha] 
	type = ParsedFunction
	value = abs(alpha_0)*(sgn_0-sqrt(r_0/(x^2+y^2))*cos(atan2(x,y)/2))
	vars = 'alpha_0 sgn_0 r_0'
	vals = '1.0 1.0 1.0'
	[../]
[]

[Materials]

	[./constants]
	type = GenericConstantMaterial
	prop_names = 'alpha_0 sgn_0 beta gamma r_0'
	prop_values = '1.0 1.0 1.0 1.0 1.0'
	[../]

        #Total Free Energy 
        [./f_eta]
	type = ParsedMaterial
	f_name = f_eta # Name of the function
	args = 'beta gamma eta' #the coefficients
	#constant_expressions = 'alpha'
	constat_names = 'alpha x y'#_0 sgn_0 r_0 x y
	function = 'alpha*eta+beta*eta^3+gamma*eta^5'
	[../]

[]

But it still givs me,
*** ERROR ***
Invalid function
alpha*eta+beta*eta^3+gamma*eta^5
sgn_0,r_0,beta,gamma,eta
in ParsedMaterialHelper.
Syntax error: Unknown identifier

I do belive that it has something to do with that x & y are coordinates in the mesh, but why can't I use mesh coordinates in the function?

**Reply by GiudGiud** [reply]
```
 [./f_eta]
type = ParsedMaterial
f_name = f_eta # Name of the function
args = 'beta gamma eta' #the coefficients
#constant_expressions = 'alpha'
constat_names = 'alpha x y'#_0 sgn_0 r_0 x y
function = 'alpha*eta+beta*eta^3+gamma*eta^5'
[../]
```

constant_names is mispelled. 
x and y are no longer in the expression for the parsed material. They are in the parsed function

**Reply by MScKarl** [reply]
Thank you for you answer, that solved the problem with the invalid function. :) 
Although, in my ParsedMaterial (I belive) , it says that:
```
*** ERROR ***
The parameter vectors constant_names and constant_values must have equal length.
```
but what constant_values is it referring to, the coupled variables are defined as AuxVariables? 'alpha' is a function calculated and referred to in materials, I do not want to give it a value, that would defeat the purpose of the function.

```
[AuxVariables]

	[./alpha_0]
        order = CONSTANT
        family = MONOMIAL
	[../]
	
	[./sgn_0]
	order = CONSTANT
	family = MONOMIAL
	[../]

	[./beta]
        order = CONSTANT
        family = MONOMIAL
	[../]

	[./gamma]
        order = CONSTANT
        family = MONOMIAL
	[../]

	[./r_0] 
        order = CONSTANT
        family = MONOMIAL
	[../]
[]
```

**Reply by GiudGiud** [reply]
alpha should not be in constant_names, it's a function.

Looks like we are missing support for functions in parsed materials. 
Let's do this instead, put it in a variable,
```
 [f_eta]
  type = ParsedMaterial
  f_name = f_eta # Name of the function
  args = 'beta gamma eta alpha' #the coefficients
  function = 'alpha*eta+beta*eta^3+gamma*eta^5'
 []

[AuxVariables]
  [alpha]
        order = CONSTANT
        family = MONOMIAL
  []
[]

[AuxKernels]
  [fill_alpha]
    type = FunctionAux
    variable = alpha
    function = alpha_fun
  []
[]

[Functions]
  [alpha] 
    type = ParsedFunction
    value = abs(alpha_0)*(sgn_0-sqrt(r_0/(x^2+y^2))*cos(atan2(x,y)/2))
    vars = 'alpha_0 sgn_0 r_0'
    vals = '1.0 1.0 1.0'
  []
[]
```

btw, make sure to use triple back-quotes when pasting code or input files. It looks a lot nicer

**Reply by MScKarl** [reply]
Thank you for your answer. I have now tried this solution and, well it runs, but when trying from what I can see MOOSE doesn't use the applied equation. Therefore, my question is why? By varying the values of the different coefficients, the outcome should vary, but it doesn’t. The impression is that some other bulk energy density potential is being used, which we haven't defined.

**Reply by GiudGiud** [reply]
What is the name of the bulk energy density potential in the code?
You defined a property with the name `f_eta`. Is that the one?

what does
```
[Debug]
  show_material_props = true
[]
```
say ?

**Reply by MScKarl** [reply]
Precisely, no matter what values are inserted into f_eta, the result is the same, which it shouldn’t.
I am a bit unsure where the results from the Debug part are showing, but at the beginning of the run this is what’s showing:


Active Materials:
    Subdomain:  (0)
      Material Name:       constants
      Property Names:      "beta" "gamma"
      Material Name:       f_eta
      Property Names:      "f_eta"

    Active Face Materials:
    Subdomain:  (0)
      Material Name:       constants_face
      Property Names:      "beta" "gamma"
      Material Name:       f_eta_face
      Property Names:      "f_eta"

Active Neighboring Materials:
    Subdomain:  (0)
      Material Name:       constants_neighbor
      Property Names:      "beta" "gamma"
      Material Name:       f_eta_neighbor
      Property Names:      "f_eta"

Active Boundary Materials:

Framework Information:
MOOSE Version:           git commit 7dee824bec on 2022-01-28
LibMesh Version:
PETSc Version:           3.15.1
SLEPc Version:           3.15.1
Current Time:            Thu May  5 11:14:16 2022
Executable Timestamp:    Sat Mar 19 16:03:00 2022

Parallelism:
  Num Processors:          12
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:
    Total:                 10201
    Local:                 892
    Min/Max/Avg:           779/896/850
  Elems:
    Total:                 10000
    Local:                 831
    Min/Max/Avg:           813/846/833
  Num Subdomains:          1
  Num Partitions:          12
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                10201
  Num Local DOFs:          892
  Variables:               "eta"
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "FIRST"

Auxiliary System:
  Num DOFs:                30000
  Num Local DOFs:          2493
  Variables:               { "beta" "gamma" "alpha" }
  Finite Element Types:    "MONOMIAL"
  Approximation Orders:    "CONSTANT"

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             Preconditioned JFNK


Time Step 0, time = 0


Consumed Material Properties:


Time Step 1, time = 0.1, dt = 0.1

**Reply by GiudGiud** [reply]
Are you specifying to use `f_eta` somewhere in the simulation?

**Reply by MScKarl** [reply]
This is how the code looks:

```
[Mesh]
    type = GeneratedMesh
    dim = 2
    nx = 100
    ny = 100
    xmin = -1
    xmax = 1
    ymin = -1
    ymax = 1
    elem_type = QUAD4
[]

[AuxVariables]

  [./beta]
    order = CONSTANT
    family = MONOMIAL
  [../]

  [./gamma]
    order = CONSTANT
    family = MONOMIAL
  [../]

  [./alpha]
    order = CONSTANT
    family = MONOMIAL
  [../]

[]

[Variables]

  [./eta]
   [./InitialCondition]
     type = RandomIC
     min = 0.001
     max = 1
   [../]
  [../]

[]

[Functions]
  [./alpha_fun] 
    type = ParsedFunction 
    value = abs(alpha_0)*(sgn_0-sqrt(r_0/(x^2+y^2))*cos(atan2(x,y)/2))
    vars = 'alpha_0 sgn_0 r_0'
    vals = '1.0 1.0 1.0'
  [../]
[]

[BCs]

  [./left]
    type = NeumannBC
    variable = 'eta'
    boundary = 'left'
    value = 0.0
  [../]

  [./right]
    type = NeumannBC
    variable = 'eta'
    boundary = 'right'
    value = 0.0
  [../]

  [./top]
    type = NeumannBC
    variable = 'eta'
    boundary = 'top'
    value = 0.0
  [../]

  [./bottom]
    type = NeumannBC
    variable = 'eta'
    boundary = 'bottom'
    value = 0.0
  [../]

[]

[AuxKernels]
  [./fill_alpha]
    type = FunctionAux
    variable = alpha
    function = alpha_fun
  [../]
[]

[Kernels]
      
  [./deta_dt]
    type = TimeDerivative
    variable = eta
  [../]

  [./ACInterface]
    type = ACInterface
    variable = eta
    kappa_name = 0.001 #prova med [1 5 7 10 50 70]*10^-3 (g)
    mob_name = 1.0 
  [../]

[]

[Materials]

  [./constants]
    type = GenericConstantMaterial
    prop_names = 'beta gamma'
    prop_values = '1.0 1.0'
  [../]

  [./f_eta]
    type = ParsedMaterial
    f_name = f_eta 
    args = 'beta gamma eta alpha' 
    function = 'alpha*eta+beta*eta^3+gamma*eta^5'
  [../]

[]

[Debug]
  show_material_props = true
[]

[Executioner]

    type = Transient
    solve_type = PJFNK 
 
    num_steps = 100
    dt = 0.1
[]

[Outputs]
    exodus = true
[]

```

**Reply by GiudGiud** [reply]
You are creating this material property ` f_eta` but you are not using it anywhere.
There are no kernels, no materials that say: `some_param_name = f_eta`

So it's not used and doesnt influence the simulation. 
Where should f_eta come in in your equation?

**Reply by MScKarl** [reply]
I apologize that it took me this long to reply. I just wanted to thank you for your answers and your patients. Finally, I got the code to run.

**Reply by biaogxb** [reply]
I want to know how you modified your code. Thank you
